### PR TITLE
feature: support pcre2

### PR DESCRIPTION
**Pull request by swananan**
I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

For openresty to support PCRE2, as I known, the following tasks need to be done:
1. lua-nginx-module supports PCRE2
2. lua-resty-core supports PCRE2 and modifies related test cases
3. stream-lua-nginx-module supports PCRE2 (It is basically the same as what lua-nginx-module needs to do)
4. openresty support PCRE2 compile option

So far I've done the first three tasks and run through the test cases using the Sanitizer version. Compared with PCRE, PCRE2 interface has changed greatly, and there are a lot workarounds in the openresty PCRE implemention, please help to review this pr, thanks.

Related PRs: 
https://github.com/openresty/lua-resty-core/pull/439
https://github.com/openresty/test-nginx/pull/162
https://github.com/openresty/stream-lua-nginx-module/pull/322

### CODE REVIEW

**Comment by swananan** [comment]
resolve https://github.com/openresty/lua-nginx-module/issues/1984

**Comment by swananan** [comment]
> @swananan Compilation error occurred when compiling nginx source code. modules/ngx_http_lua_module/src/ngx_http_lua_regex.c: In function ‘ngx_http_lua_ffi_compile_regex’: modules/ngx_http_lua_module/src/ngx_http_lua_regex.c:124:17: error: ‘re’ may be used uninitialized [-Werror=maybe-uninitialized] 124 | if (re && re->regex) { | ~~^~~~~~~ modules/ngx_http_lua_module/src/ngx_http_lua_regex.c:541:30: note: ‘re’ was declared here 541 | ngx_http_lua_regex_t *re; | ^~

Thanks for the reminder, fixed it!

**Comment by splitice** [comment]
Read throught the commit.

Only comment is perhaps https://github.com/openresty/lua-nginx-module/pull/2221/files#diff-b6199ee7c6803f9848ab6b988cc744894e299c7df3d89e53e9e09162156a0f1aR413 should say successful not 1 since it doesnt actually use the return value.

File: src/ngx_http_lua_regex.c [hunk 1]
```diff
@@ -214,8 +383,149 @@ ngx_http_lua_ffi_set_jit_stack_size(int size, u_char *errstr,
...
+            ngx_log_error(NGX_LOG_INFO, ngx_cycle->log, 0,
+                          "pcre2_jit_compile() failed: %d in \"%V\", "
+                          "ignored",
+                          ret, &re_comp->pattern);
+
+#if (NGX_DEBUG)
+
+        } else {
+            /* pcre2 jit compiled successfully */
+            ngx_log_debug0(NGX_LOG_DEBUG_HTTP, ngx_cycle->log, 0,
```
Comment (splitice) [inline src/ngx_http_lua_regex.c:413]: @swananan Here, I'd change the message

Appologies it seems I can't link to this line directly.

Comment (swananan) [reply src/ngx_http_lua_regex.c:413]: > @swananan Here, I'd change the message
> 
> Appologies it seems I can't link to this line directly.

Oh, I understand now. I kept this log solely for passing the old pcre jit related test cases. I will correct this log and make the necessary modifications to the related test cases. Thanks for your comment!

File: src/ngx_http_lua_regex.c [hunk 2]
```diff
@@ -391,6 +677,104 @@ ngx_http_lua_ffi_compile_regex(const unsigned char *pat, size_t pat_len,
...
+    if (flags & NGX_LUA_RE_NO_UTF8_CHECK) {
+        exec_opts = PCRE2_NO_UTF_CHECK;
+
+    } else {
+        exec_opts = 0;
+    }
+
+    if (flags & NGX_LUA_RE_MODE_DFA) {
+
+        int ws[NGX_LUA_RE_DFA_MODE_WORKSPACE_COUNT];
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_lua_regex.c:730]:
```suggestion
        int ws[NGX_LUA_RE_DFA_MODE_WORKSPACE_COUNT];

```

File: src/ngx_http_lua_regex.c [hunk 3]
```diff
@@ -391,6 +677,104 @@ ngx_http_lua_ffi_compile_regex(const unsigned char *pat, size_t pat_len,
...
+        rc = pcre2_dfa_match(re->regex, s, len, pos, exec_opts,
+                             ngx_regex_match_data, ngx_regex_match_context,
+                             ws, sizeof(ws)/sizeof(ws[0]));
```
Comment (zhuizhuhaomeng) [inline src/ngx_http_lua_regex.c:733]:
```suggestion
                             ws, sizeof(ws) / sizeof(ws[0]));
```

**Review by zhuizhuhaomeng — CHANGES_REQUESTED** [review]

**Comment by splitice** [comment]
I look forward to this merging. One more step forward thanks to the efforts of @swananan

**Comment by zhuizhuhaomeng** [comment]
I have merged this PR from cmdline.

**Pull request closed without merging**